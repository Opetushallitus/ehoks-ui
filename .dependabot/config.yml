version: 1
update_configs:
  - package_manager: "javascript"
    directory: "/"
    update_schedule: "monthly"
    version_requirement_updates: "increase_versions"
    commit_message:
      prefix: "Update"
      prefix_development: "Update(devDeps)" # To get separate for dev https://github.com/dependabot/feedback/issues/433
      include_scope: true
    allowed_updates:
      - match:
          dependency_type: "production"
          update_type: "security"
    default_reviewers:
      - "j-kallunki"
    default_labels:
      - "dependencies"
      - "dependabot"
