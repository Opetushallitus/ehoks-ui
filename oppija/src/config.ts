const defaultConfig = () => ({
  backendUrl: `/ehoks-oppija-backend/api/v1`
})

const config = (() => {
  const appEl = document.getElementById("app")

  if (appEl == null) {
    return defaultConfig()
  }

  const data = appEl.getAttribute("data-app-boot-config")

  if (data == null || data === "APP-BOOT-CONFIG-DEFAULT") {
    return defaultConfig()
  }

  return JSON.parse(data)
})()

if (/\bdebug=1\b/.test(window.location.search)) {
  console.log("Config: ", config)
}

const backendUrl = config.backendUrl

export const apiUrl = (path: string) => `${backendUrl}/${path}`
export const apiPrefix = "oppija"
