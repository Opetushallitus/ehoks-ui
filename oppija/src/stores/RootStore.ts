// import kebabCase from "lodash.kebabcase"
import { getEnv, Instance, types } from "mobx-state-tree"
import { EnvironmentStore } from "stores/EnvironmentStore"
import { HOKSStore } from "stores/HOKSStore"
// TODO: remove KoulutuksenJarjestajaStore when LuoHOKS route is removed
import { KoulutuksenJarjestajaStore } from "stores/KoulutuksenJarjestajaStore"
import { OppilasStore } from "stores/OppilasStore"
import { SessionStore } from "stores/SessionStore"
import { TranslationStore } from "stores/TranslationStore"
import { TyopaikanToimijaStore } from "stores/TyopaikanToimijaStore"

export interface InjectedStores {
  store: IRootStore
}

const RootStoreModel = {
  environment: types.optional(EnvironmentStore, {}),
  hoks: types.optional(HOKSStore, {}),
  isLoading: false,
  koulutuksenJarjestaja: types.optional(KoulutuksenJarjestajaStore, {}),
  oppilas: types.optional(OppilasStore, {}),
  session: types.optional(SessionStore, {}),
  translations: types.optional(TranslationStore, {}),
  tyopaikanToimija: types.optional(TyopaikanToimijaStore, {})
}

export const RootStore = types
  .model("RootStore", RootStoreModel)
  .views(self => {
    return {
      get errors() {
        return getEnv(self).errors
      }
    }
  })

export interface IRootStore extends Instance<typeof RootStore> {}
