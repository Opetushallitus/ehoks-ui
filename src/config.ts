export const apiUrl = (path: string) => `http://localhost:3000/ehoks/api/v1/${path}`
