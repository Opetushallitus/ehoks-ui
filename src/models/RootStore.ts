import camelCase from "lodash.camelcase"
// import kebabCase from "lodash.kebabcase"
import mapObj from "map-obj"
import { flow, getEnv, types } from "mobx-state-tree"
import { EducationProviderStore } from "models/EducationProviderStore"
import { StudentStore } from "models/StudentStore"
import { WorkplaceProviderStore } from "models/WorkPlaceProviderStore"
import { LearningPeriod } from "./LearningPeriod"

const RootStoreModel = {
  activeLocale: types.optional(
    types.union(types.literal("fi"), types.literal("sv")),
    "fi"
  ),
  education: types.optional(EducationProviderStore, { info: {} }),
  isLoading: false,
  learningPeriods: types.optional(types.array(LearningPeriod), []),
  student: types.optional(StudentStore, { info: {} }),
  work: types.optional(WorkplaceProviderStore, { info: {} })
}

export const RootStore = types
  .model("RootStore", RootStoreModel)
  .views(self => ({
    // views defined here should be used from other stores using: getRoot(self)
    get fetch() {
      return getEnv(self).fetch
    }
  }))
  .actions(self => {
    const fetchSingle = flow(function*(url: string) {
      const response = yield self.fetch(url)
      const model = response.data && response.data[0] ? response.data[0] : {}
      // camelCase kebab-cased object keys recursively so we can use dot syntax for accessing values
      return mapObj(
        model,
        (key: string, value: any) => [camelCase(key), value],
        {
          deep: true
        }
      )
    })

    const fetchCollection = flow(function*(url: string) {
      const response = yield self.fetch(url)
      console.log("fetchCollection", response)
      const model = response.data ? response.data : []
      return model
    })

    return { fetchSingle, fetchCollection }
  })
