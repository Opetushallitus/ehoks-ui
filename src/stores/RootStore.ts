// import kebabCase from "lodash.kebabcase"
import { getEnv, Instance, types } from "mobx-state-tree"
import { EducationProviderStore } from "stores/EducationProviderStore"
import { EnvironmentStore } from "stores/EnvironmentStore"
import { OppilasStore } from "stores/OppilasStore"
import { SessionStore } from "stores/SessionStore"
import { TranslationStore } from "stores/TranslationStore"
import { TyopaikanToimijaStore } from "stores/TyopaikanToimijaStore"

export interface ApiResponse<T> {
  meta: {
    [name: string]: any
  }
  data: T
}

export interface InjectedStores {
  store: IRootStore
}

const RootStoreModel = {
  education: types.optional(EducationProviderStore, { info: {} }),
  environment: types.optional(EnvironmentStore, {}),
  isLoading: false,
  oppilas: types.optional(OppilasStore, {}),
  session: types.optional(SessionStore, {}),
  translations: types.optional(TranslationStore, {}),
  tyopaikanToimija: types.optional(TyopaikanToimijaStore, {})
}

export const RootStore = types
  .model("RootStore", RootStoreModel)
  .views(self => {
    return {
      get errors() {
        return getEnv(self).errors
      }
    }
  })

export interface IRootStore extends Instance<typeof RootStore> {}
