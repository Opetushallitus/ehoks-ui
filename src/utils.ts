import { apiUrl } from "config"
import { Instance } from "mobx-state-tree"
import { EnvironmentStore } from "stores/EnvironmentStore"
import { InjectedStores, RootStore } from "stores/RootStore"
import { TranslationStore } from "stores/TranslationStore"
import "whatwg-fetch" // polyfill window.fetch for IE 11

// mocked fetch using local json files
export const mockFetch = (version = 0) => (url: string) => {
  const [, path] = url.split(apiUrl(""))
  return Promise.resolve({
    json: () => {
      const json = require("./stores/mocks/" +
        path.replace(/\/|\-/g, "_") +
        version +
        ".json")
      return Promise.resolve(json)
    },
    ok: true,
    statusText: "Error"
  })
}

// fetch that includes credentials
export const fetch = (url: string | Request, init: RequestInit) =>
  window.fetch(url, { credentials: "include", ...init })

export const injectSession = (stores: InjectedStores) => ({
  session: stores.store.session
})

export const getTranslations = (
  root: Instance<typeof RootStore>
): Instance<typeof TranslationStore> => {
  return root.translations
}

export const getEnvironment = (
  root: Instance<typeof RootStore>
): Instance<typeof EnvironmentStore> => {
  return root.environment
}
