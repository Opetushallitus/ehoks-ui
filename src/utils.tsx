import React from "react"
import { IntlProvider } from "react-intl"
import renderer from "react-test-renderer"
import { translations } from "./translations"

// mocked fetch using local json files
export const mockFetch = (url: string) => {
  const [, path] = url.split("http://localhost:3000/")
  return Promise.resolve(
    require("./models/mocks/" + path.replace(/\//g, "_") + ".json")
  )
}

// fetch that returns the JSON directly
export const fetch = (url: string) =>
  window.fetch(url).then(response => response.json())

function defaultCreateNodeMock(): null {
  return null
}

export const createComponentWithIntl = (
  children: React.ReactNode,
  createNodeMock?: () => any,
  props = { locale: "fi" }
) => {
  const options = { createNodeMock: createNodeMock || defaultCreateNodeMock }
  return renderer.create(
    <IntlProvider {...props} messages={translations[props.locale]}>
      {children}
    </IntlProvider>,
    options
  )
}
