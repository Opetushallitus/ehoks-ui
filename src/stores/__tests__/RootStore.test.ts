import { apiUrl } from "config"
import { mockFetch } from "../../utils"
import { ApiResponse, RootStore } from "../RootStore"

describe("RootStore", () => {
  test("constructor should produce valid tree even with empty input object", () => {
    const store = RootStore.create({}, { fetch: mockFetch() })

    expect(store.toJSON()).toEqual({
      education: {
        info: {
          basicInformation: { fi: "", sv: "" },
          hoksProcess: { fi: "", sv: "" }
        },
        isLoading: false
      },
      environment: {
        eperusteetPerusteUrl: "",
        error: "",
        isLoading: false,
        opintopolkuLoginUrl: "",
        opintopolkuLogoutUrl: ""
      },
      errors: {
        errors: []
      },
      isLoading: false,
      oppilas: {
        isLoading: false,
        perusteet: []
      },
      session: {
        error: "",
        isLoading: false,
        user: null
      },
      translations: { activeLocale: "fi", isLoading: false, translations: [] },
      work: {
        info: {
          basicInformation: { fi: "", sv: "" },
          hoksProcess: { fi: "", sv: "" }
        },
        isLoading: false
      }
    })
  })

  test("fetchSingle camelCases object keys", async () => {
    // uses session_opintopolku_1.json mock
    const store = RootStore.create({}, { fetch: mockFetch(1) })

    const result: ApiResponse<any> = await store.fetchSingle(apiUrl("session"))
    expect(result).toEqual({
      data: {
        commonName: "Teuvo",
        firstName: "Teuvo Taavetti",
        surname: "Testaaja"
      },
      meta: {
        opintopolkuLoginUrl: "http://localhost:3000/auth-dev/opintopolku-login/"
      }
    })
  })

  test("fetchCollection camelCases object keys", async () => {
    // uses lokalisointi0.json mock
    const store = RootStore.create({}, { fetch: mockFetch() })

    const result: ApiResponse<any> = await store.fetchCollection(
      apiUrl("lokalisointi")
    )
    expect(result).toEqual({
      data: [
        {
          category: "ehoks",
          createdBy: "1.2.246.562.24.31103582397",
          key: "testiavain",
          locale: "fi",
          modifiedBy: "1.2.246.562.24.31103582397",
          value: "joo"
        },
        {
          category: "ehoks",
          createdBy: "1.2.246.562.24.43953048723",
          key: "toinentesti",
          locale: "fi",
          modifiedBy: "1.2.246.562.24.43953048723",
          value: "Upeeta"
        }
      ],
      meta: {}
    })
  })
})
