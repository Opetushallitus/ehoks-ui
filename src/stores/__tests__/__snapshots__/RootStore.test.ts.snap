// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RootStore constructor should produce valid tree even with empty input object 1`] = `
Object {
  "environment": Object {
    "eperusteetPerusteUrl": "",
    "error": "",
    "isLoading": false,
    "opintopolkuLoginUrl": "",
    "opintopolkuLogoutUrl": "",
  },
  "isLoading": false,
  "koulutuksenJarjestaja": Object {
    "search": Object {
      "activePage": 0,
      "approvedOnly": false,
      "isLoading": false,
      "perPage": 10,
      "results": Array [],
      "searchText": "",
      "sortBy": "nimi",
      "sortDirection": "desc",
    },
  },
  "oppilas": Object {
    "isLoading": false,
    "perusteet": Array [],
  },
  "session": Object {
    "error": "",
    "isLoading": false,
    "user": null,
  },
  "translations": Object {
    "activeLocale": "fi",
    "isLoading": false,
    "translations": Array [],
  },
  "tyopaikanToimija": Object {
    "isLoading": false,
    "oppijat": Array [],
  },
}
`;
